### ISSUE TITLE
Provide complete implementation for table colocation

### ISSUE THREAD

**Issue by ozgune** [citusdata/citus#695]
We currently enable colocation in an indirect manner: if the user picks the same distribution column, shard count, and replication factor, the tables end up being co-located. We should look to provide a complete implementation for table colocation that involves the following sub-issues:
1. Establish relationship between two tables (in metadata) to colocate two or more tables
2. Have our shard rebalancer honor the colocation relationship defined in our metadata
3. Our APIs for table and shard creation should enable table colocation in an explicit and intuitive way (#694)
4. Foreign key constrains: Once we enforce table colocation, we can then also pass down foreign key constraints for colocated tables. (#698)
5. Once we define table colocation, we could also use the relationship between tables to simplify `INSERT INTO (SELECT ...)` for distributed rollups (#508)

We also have several checks in our logical/physical planner that checks if two tables are co-located (including subquery pushdowns). We could remove those sections to make sure we don't take additional technical debt and that the code runs faster (#730). We consider this cleanup optional and not part of the cleanup task.

**Comment by onderkalaci** [comment]
>  if the user picks the same distribution column, shard count, and replication factor, the tables end up being co-located

Though we don't end-up frequently, the shard creation logic skips a worker if it cannot create a shard, and [tries](https://github.com/citusdata/citus/blob/010cbf16fc3908bd6aa7cca69bd0c26999bf16a2/src/backend/distributed/master/master_stage_protocol.c#L387) the next worker. So, it is not always guaranteed that tables are colocated.

**Comment by onderkalaci** [comment]
Below are the some of the high-level approaches we could implement:

# **Approach 1: Infer collocation**
Do not ask user about table collocation explicitly. Instead, infer it from shard count, replication factor (and possibly some other configuration such as partition column type and worker nodes etc.)
A form of the approach discussed [here](https://docs.google.com/document/d/1QTUMUS8SDT6FnfOUfjUTtN793DXPakZM3J9jIis2pzs/edit).

**Pros:**

* More user friendly since it does not put the burden of table collocation to the user.

**Cons:**

* Becomes harder to maintain with shard rebalancer. It leads to different questions such as should a user be able to rebalance only few of the tables etc. 
  * @ozgune 's reply: I agree that this makes shard rebalancing more rigid. That said, the way this rigidity would impact the user experience will potentially be around the rebalancing performance. If we broke that down by use-case, do we have the following? If we only have the following workloads, can we then say that rebalancing performance won't be an issue for the user?

   1. Multi-tenant applications: In this case, we need to move most (or all) tables' shard placements together. Therefore, performance impact will be minimal.
   2.  Shard on user_id: In this case, we either have a single table, two co-located tables (user and events), or multiple co-located tables and small tables. My guess is most of the bigger tables will be the colocated tables. If that's the case, the observed performance impact will be low.
   3. Shard on time: We don't have colocation or colocated joins in this scenario.
   4. Data warehousing: In this case, we'll have multiple large tables sharded on different columns and there will be visible impact on rebalancing performance. However, we currently aren't targeting data warehousing workloads. Further, existing solutions seem to take the cluster offline during rebalancing and users seem to be fine with that.
* Upgrades may lead to subtle problems with the existing clusters such as different tables having the same shard count and replication factor might have the same collocation for the placements. This could be difficult to communicate with the users.
* Implementing rebalancing different pairs of the collocated tables become harder. For instance, given 4 tables, the user marks only 2 of them are collocated.
* Makes harder creating foreign key constraints since we’re not sure whether the tables are collocated while creating the tables.

  * Note that currently Citus does not allow the following alter table statement, which makes foreign key creation avaliable only on table creation:
  * ALTER TABLE public.table8 ADD CONSTRAINT table8_id_fkey FOREIGN KEY (id) REFERENCES public.table7(id)
  * Do we need to hide shard count from the user to make it simpler?
    * I think it is difficult (or not natural) with our current approach where we create distributed tables and shards in separate steps. We don't necessarily need to hide the shard count. Instead, we should create the distributed table and shards at the same step so that it becomes easier (and more natural) to infer the collocation of the tables.



# **Approach 2: Be explicit about collocation on the table creation**
**Approach 2(a)** 
Change the user API to something like the following:
```SQL
SELECT master_create_hash_distributed_table(‘new_table’, basetable = ‘old_table’)
```
`basetable` parameter could be empty, which indicates that the table is not collocated with any of the tables.
**Approach 2(b):**
Change the user API to something like:
```SQL
SELECT master_create_collocated_table(‘old_table’, ‘new_table’)
```
**Pros:**
* Simpler to implement.
* User can associate specific tables as being collocated. 

**Cons:**
* User has to decide on table collocation while creating the distributed tables. The user may not know/ be interested in the concept of collocation while creating the distributed tables.

# **Approach 3: Be explicit about table collocation when needed**
* Do not infer and do not ask about table collocation while creating the distributed table.
* Do not change the table creation APIs. 
* Instead, have another UDF which can later on associates two (or more) different tables as being collocated:
```SQL
SELECT master_collocate_tables(‘table1’, ‘table2’)
SELECT master_collocate_tables([table1, table2, ... , tableN])
```
**Pros:**
* User can associate specific tables as being collocated. 
* User does not need to define the relationship on the table creation.

**Cons:**
* What happens if the tables are not actually collocated? 
  * Error out on Citus
  * Actually collocate tables (may require shard rebalancer)
    * Means more work
* The burden is on the user while defining collocation relationship among tables.
* Once the tables are rebalanced, it is hard to define new tables as being collocated without rebalancing capabilities.

**Comment by onderkalaci** [comment]
In summary, colocation would look like the following:
* Table colocation will be represented as a new column on pg_dist_partition, called `colocationid`
* Tables which have co-located placements are given the same colocation id.
* There'll a be new sequence, called pg_dist_colocation_id_seq, to generate the new colocation ids when necessary.
* All non-hash partitioned tables and tables without any shards will have INVALID_COLOCATION_ID which is 0. This would mean that do not apply colocation for those tables, and, they would basically work as pre 5.3.


**Upgrade path:**
* This is the [README](https://gist.github.com/onderkalaci/b6547eb54cc68f86f08fbf1a54ddec79) file that includes the high-level and low-level functions that are applicable. * Note that, the user would be able to dry run the actual upgrade function as described [here](https://gist.github.com/onderkalaci/b6547eb54cc68f86f08fbf1a54ddec79#file-readme-sql-L21) (which is the default). Basically, the user would be able to see psudo_ids instead of the actual ids, but, the output which includes all the colocated tables would be the same.
* [This](https://gist.github.com/onderkalaci/b6547eb54cc68f86f08fbf1a54ddec79#file-update_table_colocation-sql) is the file that includes the upgrade functions. The user simply needs to copy & paste the file to psql.
* Another important API function is [tables_have_colocated_placements](https://gist.github.com/onderkalaci/b6547eb54cc68f86f08fbf1a54ddec79#file-readme-sql-L20)(src_table, dest_table). [Here](https://gist.github.com/onderkalaci/b6547eb54cc68f86f08fbf1a54ddec79#file-readme-sql-L159) is an example usage.
* There are good number of [tests](https://gist.github.com/onderkalaci/b6547eb54cc68f86f08fbf1a54ddec79#file-readme-sql-L24) which you can check here
* Missing parts: I probably need to do some more testing and add some extra checks such that some UDFs should be called only for hash-partitioned tables, which could be called for non-hash partitioned tables for now.


I plan to leave this process manual (i.e., not part of upgrade script). In other words, when a user upgrades to Citus 5.3, he'll get all tables to have INVALID_COLOCATION_ID . Then, if the user wants to have the co-location features, he can call the upgrade UDF. What do you think about that @samay-sharma ?

**Comment by begriffs** [comment]
Thoughts about "Approach 1" if we choose that route. Unnecessarily colocating two tables creates rebalancing overhead, doesn't it? Also although two tables may have the same shard count, replication factor and partition column type they may be partitioned on unrelated columns. A more conservative way to auto-colocate would be to do it for only those tables whose partition columns have a foreign key constraint from one to the other or from both to the same column of another table.

**Comment by onderkalaci** [comment]
By @ozgune : In approach #1, don't we just need to check the number of shards and nodes for existing tables? If a new table has the same counts, we use a deterministic algo and create the table as colocated. If not, we mark the new table as not-colocated.

Answer:
Well, that part is (more or less) right. We also need to factor in worker counts and partition column type. For instance partition columns of INT and TEXT should not be marked as co-located. Also, if a table is created on 2 workers and other is created on 3 workers, we should not mark them as co-located. Similarly, worker count helps us on the cases where the existing users rebalance only few of the tables before the upgrade.

The actual problem comes in while upgrading the existing clusters. Below, I list a few different cases, where we explicitly require a UDF (i.e., phyiscally_colocate_tables(source, dest)) which has the ability to physically move shards and co-locate tables:

Scenario 1:
User changes IP address of a worker at a given time. Since our shard creation logic sorts workers wrt their hostname/host port, we could be in trouble. So, even if the shards are created with the same shard count, replication factor, worker node count and partition column type, they are not physically co-located.

Scenario 2:
While user creating a worker shards, a worker fails temporarily. Then, the round-robin policy slips one and the tables are not physically co-located.

Scenario 3:
The user somehow calls update_shard_placement() for a single shard of a table before the upgrade. Then, we break the colocation again. (Given that we have very advanced users, I don't think that it is not possible even if we do not suppose to support this.)

Scenario 4:
Not sure, but I'm guessing there could be some other edge cases that we need to consider.


So, in general, we could end up with the same colocation configuration (shard count, replication factor, worker count and partition column) with different physical colocation layouts. Even if we find a way of representing those physical layouts exactly (i.e., the colocation configuration keeps the full physical layout) we have yet another problem. 

The problem is that, when a user tries to create a new table where we have the same (shard count, replication factor, worker count and partition column) for different co-location layouts, which layout should we follow? I don't think we can infer that in any way.

So, in a very simple words, in order our algorithm to work smoothly we need to ensure that all tables with the same (shard count, replication factor and partition column) should be physically co-located after the extension is upgraded to 6.0. This means that, we need a sophisticated upgrade script (or a way of upgrading offline) which uses the new UDF (i.e., phyiscally_colocate_tables(source, dest)) so that all tables with the same (shard count, replication factor, worker count and partition column) can be physically co-located. Later on, our algorithm works smoothly.

**Comment by ozgune** [comment]
We talked about this in detail. I'm first summarizing our meeting notes and decision on colocation below. I'll also add an update on higher level apis in a second comment.

We first agreed that we should provide a sensible default for shard count and replication factor. With that decision in place, we talked about the three approaches outlined above and a fourth one.

Approach 1
----

In this approach, we implicitly colocate all tables together.

<pre>
CREATE TABLE users (..);
SELECT master_create_distributed_table('users', 'tenant_id', 'hash');
SELECT master_create_worker_shards('users');

CREATE TABLE events (..);
SELECT master_create_distributed_table('events', 'tenant_id', 'hash'); 
SELECT master_create_worker_shards('events');
</pre>

Approach 2
----

In this approach, we ask the user to state that they are creating a co-located table.

<pre>
CREATE TABLE users (..);
SELECT master_create_distributed_table('users', 'tenant_id', 'hash'); 
SELECT master_create_worker_shards('users'); 

CREATE TABLE events (..);
SELECT master_create_colocated_table('events', 'tenant_id', 'users');
(OR SELECT master_create_distributed_table('events', 'tenant_id', baseTable=’'users');)
</pre>

Approach 3
----

In this approach, the user creates tables as they do today. They then explicitly state that two tables are colocated. A drawback to this approach is that if the ``CREATE TABLE`` statement includes foreign key constraints, the user will need to modify their schema and run ``ALTER TABLE ADD FOREIGN KEY`` later on.

<pre>
CREATE TABLE users (..);
SELECT master_create_distributed_table('users', 'tenant_id', 'hash');
SELECT master_create_worker_shards('users');

CREATE TABLE events (..);
SELECT master_create_distributed_table('events', 'tenant_id', 'hash'); 
SELECT master_create_worker_shards('events');

SELECT master_colocate_tables(‘events’, ‘users’);
</pre>

Approach 4
---

In this approach, we implicitly colocate tables together by default. We also provide an optional argument for power users to override colocation.

<pre>
CREATE TABLE users (..);
SELECT master_create_distributed_table('users', 'tenant_id', 'hash'); 
SELECT master_create_worker_shards('users'); 
=> no other tables exists, not colocated with anything

CREATE TABLE events (..);
SELECT master_create_distributed_table('events', 'tenant_id');
=> behaviour 1), i.e. colocates on type/shardcount/replication factor
SELECT master_create_worker_shards('events'); 

CREATE TABLE events_otherdimension(..);
SELECT master_create_distributed_table(‘events_otherdimension’, 'tenant_id', colocate_with => ‘’);
=> colocate with nothing
SELECT master_create_worker_shards('events_otherdimension'); 
</pre>

Decision
---

We decided to move forward with approach 4.

We also talked about our release timeline. We agreed that we should first implement the default colocation part. If we have time before our code freeze (and we expect to), we should implement the colocate_with part.

**Comment by ozgune** [comment]
@anarazel / @onderkalaci -- I have a question on ``colocate_with``'s semantics. What do we use as the default value for ``colocate_with``?

Looking at how this option could be used, we have three scenarios:

1. Default value. In this mode, we colocate the distributed table based on column type, shard count, etc. Is the default in this case then ``colocate_with => ''`` or ``colocate_with => 'all'`` (a reserved identifier that signals to colocate with all other tables)?
2. User explicitly specifies the table to colocate with. This one's easy.
3. User asks to *not* colocate this table with others. In this mode, do we use ``colocate_with => ''`` or ``colocate_with => 'none'`` (a reserved identifier that signals to colocate with nothing)?

**Comment by ozgune** [comment]
@onderkalaci -- I have an implementation question around inferring and automatically colocating tables. One option is to create a metadata table that records the options each table was created with (shard count, replication factor, partition column type, etc.).

Another option is to look at an existing distributed table's shard placements. We'd then make sure that the new distributed table's shard placements are created in the same order as the existing distributed table. We're also planning to add placement ids into our metadata tables with this release.

How do the two options compare? What are the pros/cons of each?

**Comment by onderkalaci** [comment]
Hey @ozgune,

I've a couple follow-up questions (and I'll share my answers on that)

* With Approach 4, we allow two different co-location configuration with the same shard count/rep factor and partition type. What happens when a user creates a new distributed table while there exists two such co-location groups?
  * The safest option seems to error out saying that you have to provide colocate_with option.
  * Another option would be to pick the co-location group with has greater colocationId (which means this is more recent co-location configuration)


* With Approach 4, `create_worker_shards()` become confusing and unnecessary. I see that we already plan to remove it on #694 . **But, is it in the scope of this PR?** Let me share you an example where you'd see that `create_worker_shards()` could lead to some confusion:
  * User creates a distributed table & shards
  * User creates another distributed table (not the shards yet)
    * We already pick the co-location configuration for this new table
  * Then, user changes the default shard count value before creating the shards (and thinking that it'll affect the shard count of the new table)
  * When user calls `create_worker_shards()`, we create the shards with respect to the co-location group that we've already picked.
    * We could error out in this case?

**Comment by onderkalaci** [comment]
> @onderkalaci -- I have an implementation question around inferring and automatically colocating tables. One option is to create a metadata table that records the options each table was created with (shard count, replication factor, partition column type, etc.).

> Another option is to look at an existing distributed table's shard placements. We'd then make sure that the new distributed table's shard placements are created in the same order as the existing distributed table. We're also planning to add placement ids into our metadata tables with this release.

> How do the two options compare? What are the pros/cons of each?

@ozgune I don't have a strong preference over any of two. But, slightly inclined to include the table. Please see my notes below:

**Adding table**
* A well-defined metadata seems more obvious to the user (and me!). We might end-up multiple co-location configurations with the same parameters. Users can easily look-up those. We may even refer to those co-location ids via some error messages. 
* In the future, in case we want to add some implicit parameters to the co-location configuration (such as shard placement policy), this works better.
* Relatively simpler to implement

**Not adding table**
* Does not scale, we might need to iterate through every single distributed table, shard and shard placement before creating a new distributed table.
* Co-location configurations are ambiguous to the user.

**Comment by ozgune** [comment]
@onderkalaci -- I updated my comment on approach #4 above and I'm inlining a few notes.

>> With Approach 4, we allow two different co-location configuration with the same shard count/rep factor and partition type. What happens when a user creates a new distributed table while there exists two such co-location groups?

My take is that we have one "default" colocation config that's based on shard count, replication factor, and distribution type. When the user creates a new distributed table, we by default assume they belong to the default colocation config.

If the user then creates a new table named ``XX`` and specifies that it isn't colocated with any other table, we create a new colocation config for that table, let's say "config 2".

* If the user then another table named ``YY`` and specifies that it's colocated with with table ``XX``, we assume that table ``YY`` belongs to "config 2".
* If the user creates another table named ``WW``, we assume that it belongs to "default config".
* If the user creates another table named ``ZZ`` and specifies that it isn't colocated with any other table, we create a new colocation config for that table, let's say "config 3".

>> With Approach 4, create_worker_shards() become confusing and unnecessary. I see that we already plan to remove it on #694 . But, is it in the scope of this PR? Let me share you an example where you'd see that create_worker_shards() could lead to some confusion:

I'd say yes, #694's renewed API is probably in scope for approach 4. From a task prioritization standpoint, I'd order them like the following:

1. create_distributed_table() infers and automatically colocates tables
2. We read defaults for shard count and shard replication factor from GUC
3. create_reference_table() creates small tables
4. create_distributed_table() overrides colocation with colocate_with option (your question on colocation config above)
5. We implement logic to upgrade existing users to the new colocation apis

**Comment by ozgune** [comment]
@onderkalaci -- On implementing colocation through an additional metadata table or not, I'll defer to you guys. On one of the pros/cons, I just wanted to drop a minor note.

"A well-defined metadata seems more obvious to the user." -> If we get these apis right, ideally the user wouldn't have to know about this metadata table at all. 😃

**Comment by onderkalaci** [comment]
@ozgune We can chat about the following on the meeting, but, for now just noting the options:

>> If the user then creates a new table named ``XX`` and specifies that it isn't colocated with any other table, we create a new colocation config for that table, let's say "config 2".
>>
* If the user then another table named ``YY`` and specifies that it's colocated with with table ``XX``, we assume that table ``YY`` belongs to "config 2".
* If the user creates another table named ``WW``, we assume that it belongs to "default config".
* If the user creates another table named ``ZZ`` and specifies that it isn't colocated with any other table, we create a new colocation config for that table, let's say "config 3".

In summary, you expect that shard creation follows the `initial`  co-location configuration for the same (shard count, replication factor and part type). We can list all the options as follows:

* Follow the initial co-location configuration
* Error out and explicitly ask about `colocate_with` option. Given that user could only be in this situation if he already used `colocate_with` option, this is not much annoying. The intent here is to prevent user creating wrong/un-expected colocated tables. 
* Follow the latest co-location configuration

I was thinking that we should go with either 2 or 3. But, I'm ok with any of the above and just wanted to drop a note.

**Comment by ozgune** [comment]
@onderkalaci / @metdos, I wanted to drop a quick note. #844 is a candidate feature for our v6.1 release.

Could you spend less than fifteen minutes looking into #844? We don't need to think about this right now -- just wanted to bring it to your attention.

**Comment by ozgune** [comment]
@onderkalaci / @metdos -- Now that we have the colocation APIs in v6.0, I'm closing this issue. If you think we have open questions related to this issue, please feel free to reopen it.
